### PR TITLE
Django 1.9, 1.10, and 1.11 upgrades

### PR DESCRIPTION
**Pull request by vkurup**
Starting with #119, aim to get this working on Django 1.11. It currently only works in 1.8

### CODE REVIEW

File: caching/base.py [hunk 1]
```diff
@@ -157,6 +157,18 @@ def __setstate__(self, state):
...
+    def _fetch_all(self):
+        """
+        Django 1.11 changed _fetch_all to use self._iterable_class() rather than
+        self.iterator(). That bypasses our iterator, so override Queryset._fetch_all
+        to use our iterator.
+
+        https://github.com/django/django/commit/f3b7c059367a4e82bbfc7e4f0d42b10975e79f0c#diff-5b0dda5eb9a242c15879dc9cd2121379
+        """
+        if self._result_cache is None:
+            self._result_cache = list(self.iterator())
```
Comment (tobiasmcnulty) [inline caching/base.py]: I believe this change will cause a server-side cursor to be used for `_fetch_all` (if not globally disabled via `DISABLE_SERVER_SIDE_CURSORS`) on Django 1.11.1+ due two these lines:

- https://github.com/vkurup/django-cache-machine/blob/6a547d8e4295985b2a84d4e115b1192cff540dbc/caching/base.py#L181
- https://github.com/django/django/blob/stable/1.11.x/django/db/models/query.py#L322-L323

I think we need to make sure `_fetch_all()` still uses `self._iterable_class()` directly rather than `self.iterator()` on Django 1.11+, perhaps via something along the lines of the following:

- Separate our `iterator()` logic into a separate method, e.g., `_iterate_on(iterable)`
- In `_fetch_all()`, call `list(self._iterate_on(iter(self._iterable_class(self))))` for Django 1.11+ or `list(self._iterate_on(self.iterator()))` for older Djangos (`_iterable_class` unfortunately does not exist in Django 1.8).
- In `iterator()`, simply return `self._iterate_on(super(CachingQuerySet, self).iterator())`

Does that seem crazy? What do you think?

Comment (tobiasmcnulty) [reply caching/base.py]: Or better yet, rather than check the version directly, use duck typing as you already did here: https://github.com/django-cache-machine/django-cache-machine/pull/125/commits/24c977923ab5fee5525fb5fea0df6b4859e53316#diff-76150a51cd21d8d6b30aaaab22c7f114R179

`_fetch_all()` and `iterator()` are unchanged between Django 1.9-1.10, so it should be safe to use `self._iterable_class` directly in those versions (in addition to Django 1.11)

`_fetch_all()`:
- 1.9: https://github.com/django/django/blob/stable/1.9.x/django/db/models/query.py#L1072-L1076
- 1.10: https://github.com/django/django/blob/stable/1.10.x/django/db/models/query.py#L1085-L1089

`iterator()`:
- 1.9: https://github.com/django/django/blob/stable/1.9.x/django/db/models/query.py#L325-L330
- 1.10: https://github.com/django/django/blob/stable/1.10.x/django/db/models/query.py#L323-L328

**Comment by vkurup** [comment]
@tobiasmcnulty I made an update, trying to avoid overriding `_fetch_all`. This involved the following changes:

* Creating `CachingModelIterable` as a subclass of django's `ModelIterable`, which allows us to use the superclass's `__iter__` definition easily
* Changed the constructor of `CacheMachine` to allow it to share most of its code with `CachingModelIterable`
* Some 'conditional' hacks to make sure we load the correct iterator for Django 1.8 vs Django 1.9+

Let me know what you think!

File: caching/base.py [hunk 2]
```diff
@@ -104,28 +112,47 @@ def __iter__(self):
...
-        try:
-            while True:
-                obj = next(iterator)
-                obj.from_cache = False
-                to_cache.append(obj)
-                yield obj
-        except StopIteration:
-            if to_cache or config.CACHE_EMPTY_QUERYSETS:
-                self.cache_objects(to_cache, query_key)
-            raise
```
Comment (tobiasmcnulty) [inline caching/base.py]: This change loses the `StopIteration` re-raised here -- is that intentional?

File: caching/base.py [hunk 3]
```diff
@@ -86,16 +71,39 @@ def query_key(self):
...
+    def cache_objects(self, objects, query_key):
+        """Cache query_key => objects, then update the flush lists."""
+        log.debug('query_key: %s' % query_key)
+        query_flush = flush_key(self.queryset.query_key())
+        log.debug('query_flush: %s' % query_flush)
+        cache.add(query_key, objects, timeout=self.timeout)
+        invalidator.cache_objects(self.queryset.model, objects, query_key, query_flush)
+
     def __iter__(self):
+        if hasattr(super(CacheInternalCommonMixin, self), '__iter__'):
```
Comment (tobiasmcnulty) [inline caching/base.py]: Perhaps `try:`/`except AttributeError:` to avoid `hasattr()` usage? The Django 1.8 bit can go away eventually, and there will be no performance hit for Django 1.9+

File: caching/base.py (hunk 2 above)
Comment (vkurup) [reply caching/base.py]: Ahh... I think I missed that (I thought I was just simplifying an iteration through a iterator). I'll have to see if I can find a way to test that, or maybe just revert back to the old version of the code.

File: caching/base.py (hunk 3 above)
Comment (vkurup) [reply caching/base.py]: +1

File: caching/base.py [hunk 4]
```diff
@@ -104,28 +112,47 @@ def __iter__(self):
...
 
-    def cache_objects(self, objects, query_key):
-        """Cache query_key => objects, then update the flush lists."""
-        log.debug('query_key: %s' % query_key)
-        query_flush = flush_key(self.query_string)
-        log.debug('query_flush: %s' % query_flush)
-        cache.add(query_key, objects, timeout=self.timeout)
-        invalidator.cache_objects(self.model, objects, query_key, query_flush)
+
+class CacheMachine(CacheInternalCommonMixin):
```
Comment (tobiasmcnulty) [inline caching/base.py]: Not entirely sure this will work, but I wonder if `CacheInternalCommonMixin`, `CacheMachine`, and `CachingModelIterable` could all be collapsed into a single class (e.g., by setting `ModelIterable = object` when it fails to import). Definitely ignore this comment if that would complicate rather than simplify things!

Comment (vkurup) [reply caching/base.py]: I tried doing that, but gave up early because of issues that I can't remember. But some of those problems might have been 2 small issues in the tests which I since identified, so let me try again.

File: caching/base.py (hunk 2 above)
Comment (tobiasmcnulty) [reply caching/base.py]: Ah, perhaps it doesn't matter? https://stackoverflow.com/questions/4019971/how-to-implement-iter-self-for-a-container-object-python#comment67775147_4019987

Comment (vkurup) [reply caching/base.py]: Actually, this might be a part of Python that I just don't understand well. I'd think that I could just fix this by adding a `raise StopIteration` on line 128 (at the same indentation level as the `if` on line 126). But that seems superfluous, since it's the last line of the method and the iteration will end anyways. Am I missing something?

Comment (tobiasmcnulty) [reply caching/base.py]: https://www.python.org/dev/peps/pep-0479/#rationale

File: caching/base.py [hunk 5]
```diff
@@ -86,16 +71,39 @@ def query_key(self):
...
+            # no cache, just iterate and return the results
+            for obj in iterator():
+                yield obj
+            return
+
+        # Try to fetch from the cache.
         try:
             query_key = self.query_key()
         except query.EmptyResultSet:
             raise StopIteration
```
Comment (tobiasmcnulty) [inline caching/base.py]: Given our findings about `StopIteration` elsewhere, I think this should be converted to a simple `return`

File: caching/base.py [hunk 6]
```diff
@@ -309,11 +332,13 @@ def __iter__(self):
             while True:
                 yield next(iterator)
         else:
-            sql = self.raw_query % tuple(self.params)
-            for obj in CacheMachine(self.model, sql, iterator, timeout=self.timeout):
+            for obj in CacheMachine(self, iterator, timeout=self.timeout):
                 yield obj
             raise StopIteration
```
Comment (tobiasmcnulty) [inline caching/base.py]: I think this line can be deleted (so we get the implicit `return None` instead)

**Comment by tobiasmcnulty** [comment]
⛵️ 

Thanks again for taking this on!
